types/index.d.ts

```diff
@@ -1906,17 +1839,17 @@ declare namespace math {
      * traversed. The function must return a boolean.
      */
     filter(
-      x: Matrix | MathArray | string[],
+      x: MathCollection | string[],
```

Comment:
In general, the functions of mathjs follow the rule "Array in -> Array out, Matrix in -> Matrix out". For functions with a single argument we could utilize that using generics. For functions that have two arguments it is more complicated (if one of the two is a Matrix, the output will be a Matrix, if both are an Array, the output is an Array). 

Maybe not for this PR, but it would be nice to utilize that knowledge. What do you think?

Reply:
yeah, I agree. Some functions like dotDivide and divide give better Type restrictions following this reasoning. I can make another PR refactoring such functions :)

Reply:
Nice! But indeed one step at a time :)